### PR TITLE
[move-ide] Fixed auto-completion on the dot character

### PR DESCRIPTION
**Pull request by awelc**
## Description 

This PR fixes a certain problem with auto-completion on the `.` character. Previously, once a part of the access chain after the `.` was parsed, auto-completion information for the `.` character location was no longer generated. This is particularly problematic if someone expects auto-completion in the middle of the function code, for example:
```move
public struct SomeStruct has drop {
    some_field: u64,
}

public fun foo(_param: SomeStruct) {}

public fun bar(): SomeStruct {
    let val = SomeStruct{some_field: 42};
    val.
    SomeStruct {some_field: 42}
}
```
It is expected that auto-completion would work correctly after `val.` but previously it did not because this expression was actually parsed as `val.SomeStruct`. This PR fixes this:
image

Previously, auto-completion for the `.` location would also not work if what follows the `.` is a function call (whether the function being called was resolved correctly or not). It now works correctly in both cases:
image
image

## Test plan 

All new and (updated) all tests must pass

### CODE REVIEW

File: external-crates/move/crates/move-compiler/src/naming/ast.rs [hunk 1]
```diff
@@ -358,7 +358,7 @@ pub type LambdaLValues = Spanned<LambdaLValues_>;
 #[derive(Debug, PartialEq, Clone)]
 pub enum ExpDotted_ {
     Exp(Box<Exp>),
-    Dot(Box<ExpDotted>, Field),
```
Comment (cgswords) [inline external-crates/move/crates/move-compiler/src/naming/ast.rs:361]: Nit: `loc` should be in the middle, with a comment indicating what the location is

File: external-crates/move/crates/move-compiler/src/expansion/ast.rs [hunk 2]
```diff
@@ -393,6 +393,7 @@ pub enum Exp_ {
         /* is_macro */ Option<Loc>,
         Option<Vec<Type>>,
         Spanned<Vec<Exp>>,
+        Loc, // location of the dot
```
Comment (cgswords) [inline external-crates/move/crates/move-compiler/src/expansion/ast.rs:396]: Put `Loc` between the boxed expression and the name.

File: external-crates/move/crates/move-compiler/src/expansion/ast.rs [hunk 3]
```diff
@@ -336,7 +336,7 @@ pub type LambdaLValues = Spanned<LambdaLValues_>;
 #[allow(clippy::large_enum_variant)]
 pub enum ExpDotted_ {
     Exp(Box<Exp>),
-    Dot(Box<ExpDotted>, Name),
+    Dot(Box<ExpDotted>, Name, Loc),
```
Comment (cgswords) [inline external-crates/move/crates/move-compiler/src/expansion/ast.rs:339]: `Loc` in the middle

File: external-crates/move/crates/move-compiler/src/typing/translate.rs [hunk 4]
```diff
@@ -3393,12 +3421,15 @@ fn resolve_exp_dotted(
...
-        result
-    } else {
-        result
     }
+    if context.env.ide_mode() {
+        if let Some(mdot_loc) = method_dot_loc {
+            ide_report_autocomplete(context, &mdot_loc, &edotted_ty);
+        }
+    }
+    result
```
Comment (cgswords) [inline external-crates/move/crates/move-compiler/src/typing/translate.rs:3432]: This check is redundant.

```suggestion
    if let Some(mdot_loc) = method_dot_loc {
        ide_report_autocomplete(context, &mdot_loc, &edotted_ty);
    }
    result
```

**Review by cgswords — APPROVED** [review]
LGTM except for one tiny nit

File: external-crates/move/crates/move-compiler/src/typing/translate.rs (hunk 4 above)
Comment (awelc) [reply external-crates/move/crates/move-compiler/src/typing/translate.rs]: It wasn't 100% redundant as there were places where `method_dot_loc` argument was set to an actual location that were not guarded by the `ide_mode` flag, but it is now after some minor additional changes to `method_call` and `macro_method_call` functions

**Pull request merged**